### ISSUE TITLE
Ligatures slow down (even freeze) Emacs built from the master branch

### ISSUE THREAD

**Issue by hexclover** [mickeynp/ligature.el#36]
Background
====

I use Arch Linux and have been using [an AUR package for Emacs master branch](https://aur.archlinux.org/packages/emacs-native-comp-git-enhanced/) and ```ligature.el``` for a few months; they worked flawlessly :-) However, recently after an update, my Emacs became really laggy and can become unresponsive when moving the cursor in a buffer containing lots of ligatures. I had to disable ```ligature-mode``` to make Emacs as smooth as before. Yesterday I had the time to do a bisect and found the commit **to Emacs** that introduced the problem. If I revert this commit and build Emacs, I can use `ligature.el` with no problem at all.

System Information
====

- Distribution: Arch Linux
- `M-x emacs-version`: `GNU Emacs 29.0.50 (build 3, x86_64-pc-linux-gnu, GTK+ Version 3.24.31, cairo version 1.17.4)
 of 2021-12-31`
- `cairo-version-string`: `1.17.4`
- `system-configuration-features`: `ACL CAIRO DBUS FREETYPE GIF GLIB GMP GNUTLS GPM GSETTINGS HARFBUZZ JPEG JSON LCMS2 LIBOTF LIBSYSTEMD LIBXML2 M17N_FLT MODULES NOTIFY INOTIFY PDUMPER PNG RSVG SECCOMP SOUND SQLITE3 THREADS TIFF TOOLKIT_SCROLL_BARS WEBP X11 XDBE XIM XPM GTK3 ZLIB`
- Fonts: tested with Fira Code 6.2 and JetBrains Mono 2.242

Ways to Reproduce
====

I clone the Emacs repository from https://github.com/emacs-mirror/emacs and directly run `make` in it to build a "vanilla" Emacs. Then load and visit this minimal init file (`emacs -q -l problematic.el problematic.el`):

```lisp
(package-initialize)
(set-face-attribute 'default nil :family "JetBrains Mono" :height 90) ;; or Fira Code, etc.
(use-package ligature
  :load-path "~/.emacs.d/site-lisp/ligature.el"
  :config
  ;; Enable the "www" ligature in every possible major mode
  (ligature-set-ligatures 't '("www"))
  ;; Enable traditional ligature support in eww-mode, if the
  ;; `variable-pitch' face supports it
  (ligature-set-ligatures 'eww-mode '("ff" "fi" "ffi"))
  ;; Enable all Cascadia Code ligatures in programming modes
  (ligature-set-ligatures 'prog-mode '("|||>" "<|||" "<==>" "<!--" "####" "~~>" "***" "||=" "||>"
                                       ":::" "::=" "=:=" "===" "==>" "=!=" "=>>" "=<<" "=/=" "!=="
                                       "!!." ">=>" ">>=" ">>>" ">>-" ">->" "->>" "-->" "---" "-<<"
                                       "<~~" "<~>" "<*>" "<||" "<|>" "<$>" "<==" "<=>" "<=<" "<->"
                                       "<--" "<-<" "<<=" "<<-" "<<<" "<+>" "</>" "###" "#_(" "..<"
                                       "..." "+++" "/==" "///" "_|_" "www" "&&" "^=" "~~" "~@" "~="
                                       "~>" "~-" "**" "*>" "*/" "||" "|}" "|]" "|=" "|>" "|-" "{|"
                                       "[|" "]#" "::" ":=" ":>" ":<" "$>" "==" "=>" "!=" "!!" ">:"
                                       ">=" ">>" ">-" "-~" "-|" "->" "--" "-<" "<~" "<*" "<|" "<:"
                                       "<$" "<=" "<>" "<-" "<<" "<+" "</" "#{" "#[" "#:" "#=" "#!"
                                       "##" "#(" "#?" "#_" "%%" ".=" ".-" ".." ".?" "+>" "++" "?:"
                                       "?=" "?." "??" ";;" "/*" "/=" "/>" "//" "__" "~~" "(*" "*)"
                                       "\\\\" "://"))
  ;; Enables ligature checks globally in all buffers. You can also do it
  ;; per mode with `ligature-mode'.
  (global-ligature-mode t))
```

The init file itself contains many ligatures in it. Make the init file the sole window and do various things to it, for example moving the cursor quickly and randomly with mouse clicks/scrolling/`C-n`/`C-p` etc. (especially moves that cause Emacs to scroll the screen) and selecting text. Then soon (maybe in 30 seconds) Emacs will become laggy and finally responsive; I have to press `C-g` and switch to another buffer to make Emacs responsive again.

Another possible and quicker way to reproduce the problem is: load the init file as usual, switch to the scratch buffer, and type `C-u 2000 =`. Then moving the cursor can completely freeze Emacs. If I use an Emacs version that does not have the problem (with ligatures enabled), the lag when moving cursor is visible but acceptable.

Bisection
====

I tracked down the problematic commit after a `git bisect` session: [`[dae3c4e89b27fa008d029a30cf7bd02fc3fbb6e2] Allow choosing regular-weighted fonts when medium-weighted exist`](https://git.savannah.gnu.org/cgit/emacs.git/commit/?id=dae3c4e89b27fa008d029a30cf7bd02fc3fbb6e2). If I built Emacs from the master branch but revert this commit, everything returns to normal (except I cannot `choose regular-weighted fonts when medium-weighted exist` as the commit message says). However, I don't see how this commit causes the problem, neither do I know how to fix it properly. And actually I am not sure if it is the Emacs commit, `ligature.el`, or my specific system configuration that is buggy...

**Comment by mickeynp** [comment]
Hm. Sounds very strange. I don't see how ligature.el is the cause as it calls into Emacs to do all the heavy lifting. ligature's really just a fancy wrapper around some composition table stuff. You can try amending the number of ligatures to see if that makes a difference also?

I recommend you `M-x report-emacs-bug` and mention your findings. In particular, ligature uses `set-char-table-range` to create the ligatures, so it's possible this is somehow causing problems in Harfbuzz or elsewhere. 

Having said all that, applying ligatures will definitely impact performance a little bit due to how it's implemented in Emacs. ligature feeds Emacs regular expressions that map to ligatures. I've no idea if ligation engines in other tools (LaTeX, browsers) also work like this, though.

**Comment by hexclover** [comment]
Thank you! It turns out that the problem is not due to `ligature.el` but some older bug in Emacs and some funny coincidences. I've submitted a patch upstream that hopefully fixes it. After all this is really strange as you said :b

**Comment by mickeynp** [comment]
I'm curious what the resolution was exactly?

**Comment by hexclover** [comment]
The root cause(?) is actually an integer size mismatch problem in `src/xfaces.c` of the Emacs source code. If you don't care reading lengthy comments, here are the details:

<details><summary>Details</summary>

Before the fix, the `face_for_font` function in it reads

```C
int
face_for_font (struct frame *f, Lisp_Object font_object,
               struct face *base_face)
{
  struct face_cache *cache = FRAME_FACE_CACHE (f);
  unsigned hash; // NB!
  int i;
  // ...
  hash = lface_hash (base_face->lface);
  i = hash % FACE_CACHE_BUCKETS_SIZE;
  // the function goes on to look up the face in hash bucket i
  // if not found, it creates a new face and store it in the cache
}
```

The problem is: the `hash` variable is declared with type `unsigned`, whereas *all* other parts of the file use `uintptr_t` for hash values, including the `cache_face` function responsible for storing the newly created face in the cache (and it does not take the value from the `hash` variable here, BTW). So if on some machine `unsigned` is shorter than `uintptr_t`, and `lface_hash` returns a value larger than `UINT_MAX` for some specific `base_face`, then the value in `hash` is truncated, and the lookup and storing will be in different buckets, so the cache becomes useless and a new face is created every time `face_for_font` is called with the specific `base_face`! The cache misses and the mounting of newly created faces can eventually freeze Emacs.

The fix is simple; just change it to use `uintptr_t` instead of `unsigned`.

The more interesting part is, *why the combination and commit `dae3c4` cause this bug to manifest itself*? First, enabling ligatures can cause *lots* of calls to `face_for_font` (I'm not sure if it is a problem by itself, but I have tried other methods to enable ligatures from https://github.com/tonsky/FiraCode/wiki/Emacs-instructions, so I don't think it's specific to `ligature.el`); otherwise the function is called for only a few times, so the cache misses are not a big problem. Second, the `lface_hash` usually does not return values greater than `UINT_MAX`, but commit `dae3c4` causes it to do so (?!) for my machine. The commit reads

```diff
-      if (numeric >= FC_WEIGHT_REGULAR && numeric < FC_WEIGHT_MEDIUM)
-	numeric = FC_WEIGHT_MEDIUM;
```

As its commit message indicates, it allows choosing the **regular** weight for a font in the presence of a medium weight. The result is, in the hash function below, the `v[LFACE_WEIGHT_INDEX]` (a Lisp symbol representing the font weight) can be the symbol `'regular`:

```C
static uintptr_t
lface_hash (Lisp_Object *v)
{
  return (hash_string_case_insensitive (v[LFACE_FAMILY_INDEX])
	  ^ hash_string_case_insensitive (v[LFACE_FOUNDRY_INDEX])
	  ^ hash_string_case_insensitive (v[LFACE_FOREGROUND_INDEX])
	  ^ hash_string_case_insensitive (v[LFACE_BACKGROUND_INDEX])
	  ^ XHASH (v[LFACE_WEIGHT_INDEX])
	  ^ XHASH (v[LFACE_SLANT_INDEX])
	  ^ XHASH (v[LFACE_SWIDTH_INDEX])
	  ^ XHASH (v[LFACE_HEIGHT_INDEX]));
}
```

The strange part is: the address of the symbol `'regular` is in a different section of memory (I don't know an exact word for it, sorry) than symbols like `'semibold`, `'medium`, `'bold`, etc. So when `XHASH`-ing these weight symbols, `'regular` becomes a very large value (much larger than `UINT_MAX`), and other symbols become rather small values. I found that the symbol `'regular` is actually located in a memory area mapped to [`emacs.pdmp`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Building-Emacs.html), so maybe a symbol with name `regular` is created somewhere in the initialization process in some basic Lisp library of Emacs. Anyway, this should not be a problem itself and is unrelated to the final fix.

</details>
